pydeequ/checks.py

```diff
@@ -116,6 +116,14 @@ def addConstraint(self, constraint):
         self.constraints.append(constraint)
         self._Check = constraint._Check
 
+    def where(self, filter: str):
+        test = self._Check.getClass()
+        if self._Check.getClass().toString().endswith("CheckWithLastConstraintFilterable"):
```

Comment:
This won't work for `Check` classes derived from `com.amazon.deequ.checks.CheckWithLastConstraintFilterable` unless their class name ends in `CheckWithLastConstraintFilterable`.  That may not pass muster with pydeequ maintainers IMHO.

also the `test` variable declared above does not seem to be used.

Here's something based on the code from py4j - https://github.com/py4j/py4j/blob/9e61991c7ac0c61066486204fe94936a1ff4e186/py4j-python/src/py4j/java_gateway.py#L445

```python
    def where(self, filter: str):
        is_filterable = self._jvm.py4j.reflection.TypeUtil.isInstanceOf(
            "com.amazon.deequ.checks.CheckWithLastConstraintFilterable", self._Check)
        if not is_filterable:
            raise TypeError(f"Expected CheckWithLastConstraintFilterable class, not {self._Check.getClass()}")
        self._Check = self._Check.where(filter)
        return self
```

The issue with this is the performance cost of `TypeUtil.isInstanceOf` which seems to use reflection for every call. It might be better to cache a copy of the class using `ReflectionUtil.classForName` and use `Class.isInstance`. However my suspicion is that this is not performance intensive code. so reflection might be just fine.

on a side note, i find it odd that this pydeequ Check object seems to behave as a mutable object. But the original scala Check seems to be an immutable value type.

Reply:
Sorry for not being more encouraging with my comment. I should have started with a nicer intro:

Great job with this open source submission. It takes a lot of research to draft a change like this.

Reply:
My concern was that if they restructure the project and the package changes this will break, but overall pydeequ is a simplified interface for deequ. A lot of parameters in the python version are hardcoded from scala and it leads to high dependency on the latter. See for example: https://github.com/awslabs/python-deequ/issues/169 
Also deleted the test variable.

Reply:
The ideal case would be that pydeequ would match the way the scala implementation works - i.e. we would have a python CheckWithLastConstraintFilterable which is a subclass of Check and only have the where method on it. It's not clear to me how scala code provides an instance of CheckWithLastConstraintFilterable. When i looked at the pydeequ tests you added, i couldn't tell how self._Check is assigned a subclass instance.

seems like we'd have to create python subclass instances (instead of return self) depending on the actual type of _Check. pseudo code would be like

if is_check_subclass:
  return newSubclassCtor(self)
return self

which is kinda overkill, considering that we're already treating this as a mutable object.